File: Singly_Circluar_LinkedList/LinkedList.ts
class NODE 
{
    data : number;
    next : NODE | null;

    constructor()
    {
        this.data = 0;
        this.next = null;
    }
}

class SinglyCL
{
    Head : NODE | any;
    Tail : NODE | any;

    constructor()
    {
        this.Head = null;
        this.Tail = null;
    }

    InsertFirst(value : number)
    {
        var newn : NODE = new NODE();
        newn.data = value;
        newn.next = null;

        if((this.Head == null)&&(this.Tail == null))
        {
            this.Head = newn;
            this.Tail = newn;
        }
        else 
        {
            newn.next = this.Head;
            this.Head = newn;
        }
        (this.Tail).next = this.Head;
    }

    InsertLast(value : number) 
    {
        var newn : NODE = new NODE();
        newn.data = value;
        newn.next = null;

        if((this.Head == null)&&(this.Tail == null))
        {
            this.Head = newn;
            this.Tail = newn;
        }
        else 
        {
            (this.Tail).next = newn;
            this.Tail = newn;
        }
        (this.Tail).next = this.Head;
    }

    InsertAtPos(value : number,ipos : number) 
    {
        var isize : number = 0;
        var i : number = 0;
        var temp : NODE | any = this.Head;
        var newn : NODE | any = null;
        isize = this.Count();

        if((ipos < 1)||(ipos > (isize + 1)))
        {
            return;
        }
        else if(ipos == 1)
        {
            this.InsertFirst(value);
        }
        else if(ipos == (isize + 1))
        {
            this.InsertLast(value);
        }
        else 
        {
            newn = new NODE();
            newn.data = value;
            newn.next = null;

            for(i = 1;i < (ipos - 1);i++)
            {
                temp = temp.next;
            }

            newn.next = temp.next;
            temp.next = newn;
        }
    }

    DeleteFirst()
    {
        if((this.Head == null) && (this.Tail == null))
        {
            return;
        }
        else if(this.Head == this.Tail)
        {
            delete(this.Head);
            this.Head = null;
            this.Tail = null;
        }
        else 
        {
            this.Head = (this.Head).next;
            delete((this.Tail).next);
            (this.Tail).next = this.Head;
        }
    }

    DeleteLast()
    {
        var temp : NODE | any = this.Head;
        if((this.Head == null)&&(this.Tail == null))
        {
            return;
        }
        else if(this.Head == this.Tail)
        {
            delete(this.Head);
            this.Head = null;
            this.Tail = null;
        }
        else 
        {
            while(temp.next != this.Tail)
            {
                temp = temp.next;
            }

            delete(this.Tail);
            this.Tail = temp;
        }
        (this.Tail).next = this.Head;
    }

    DeleteAtPos(ipos : number)
    {
        var i : number = 0;
        var isize : number = 0;
        var temp : NODE | any = this.Head;
        var temp2 : NODE | any = null;

        isize = this.Count();

        if((ipos < 1)||(ipos > isize))
        {
            return;
        }
        else if(ipos == 1)
        {
            this.DeleteFirst();
        }
        else if(ipos == isize)
        {
            this.DeleteLast();
        }
        else 
        {
            for(i = 1;i < (ipos - 1);i++)
            {
                temp = temp.next;
            }

            temp2 = temp.next;
            temp.next = temp2.next;

            temp2 = null;
        }
    }

    Display() 
    {
        var Head : NODE | any = this.Head;

        console.log("Elements of Linked List are : ");
        do 
        {
            process.stdout.write("|"+Head.data+"|<->");
            Head = Head.next
        }while(Head != (this.Tail).next);
        console.log();
    }

    Count() : number  
    {
        var Head : NODE | any = this.Head;
        var icnt : number = 0;

        do 
        {  
            icnt++; 
            Head = Head.next
        }while(Head != (this.Tail).next);

        return icnt;
    }
}

function main() 
{
    var iret : number = 0;

    console.log("Jay Ganesh...");

    var sobj = new SinglyCL();

    sobj.InsertFirst(51);
    sobj.InsertFirst(11);
    sobj.InsertFirst(10);
    sobj.InsertLast(60);
    sobj.InsertLast(65);
    sobj.InsertLast(100);
    sobj.InsertAtPos(21,3);

    sobj.Display();
    iret = sobj.Count();
    console.log("Count of Elements present in Linked List is : %d",iret);

    console.log("-------Output After Performin DeleteFirst Operation------");
    sobj.DeleteFirst();
    sobj.Display();
    iret = sobj.Count();
    console.log("Count of Elements present in Linked List is : %d",iret);

    console.log("-------Output After Performin DeleteLast Operation------");
    sobj.DeleteLast();
    sobj.Display();
    iret = sobj.Count();
    console.log("Count of Elements present in Linked List is : %d",iret);

    console.log("-------Output After Performin DeleteAtPos Operation------");
    sobj.DeleteAtPos(5);
    sobj.Display();
    iret = sobj.Count();
    console.log("Count of Elements present in Linked List is : %d",iret);
}


main();

File: DoublyCircularLinkedList/LinkedList.py
class Node :
    data = 0;
    next = None;
    prev = None;

    def __init__(self) :
        self.data  = 0;
        self.next = None;
        self.prev = None;

class DoublyCLL :
    __Head = None;
    __Tail = None;

    def __init__(self) :
        self.__Head = None;
        self.__Tail = None;
    
    def InsertFirst(self,value) :
        newn = Node();
        newn.data = value;
        newn.next = None;
        newn.prev = None;

        if((self.__Head == None)and(self.__Tail == None)) :
            self.__Head = newn;
            self.__Tail = newn;
        else :
            newn.next = self.__Head;
            (self.__Head).prev = newn;
            self.__Head = newn;
        (self.__Head).prev = self.__Tail;
        (self.__Tail).next = self.__Head;

    def InsertLast(self,value) :
        newn = Node();
        newn.data = value;
        newn.next = None;
        newn.prev = None;

        if((self.__Head == None)and(self.__Tail == None)) :
           self.__Head = newn;
           self.__Tail = newn;
        else :
            (self.__Tail).next = newn;
            newn.prev = self.__Tail;
            self.__Tail = newn;
        (self.__Tail).next = self.__Head;
        (self.__Head).prev = self.__Tail;  

    def InsertAtPos(self,value,ipos) :
        i = 0;
        isize = 0;
        newn = None;
        temp = self.__Head;

        isize = self.Count();

        if((ipos < 1)or(ipos > (isize + 1))) :
            return;
        elif(ipos == 1) :
            self.InsertFirst(value);
        elif(ipos == (isize + 1)) :
            self.InsertLast(value);
        else :
            newn = Node();
            newn.data = value;
            newn.next = None;
            newn.prev = None;

            for i in range(1,(ipos - 1)) :
                temp = temp.next;
            
            newn.next = temp.next;
            newn.prev = temp;
            (temp.next).prev = newn;
            temp.next = newn;

    def Display(self) :
        Head = self.__Head;
        Tail = self.__Tail;

        if((Head == None)and(Tail == None)) :
            return;
        else :
            print("Elements of Linked List are : ");
            while(Head != Tail) :                
                print("|%d|<=>"%Head.data,end = "");
                Head = Head.next;
            print("|%d|<=>"%Head.data,end = "");
            print();

    def DeleteFirst(self) :
        if((self.__Head == None)and(self.__Tail == None)) :
            return;
        elif(self.__Head == self.__Tail) :
            del(self.__Head);
            self.__Head = None;
            self.__Tail = None;
        else :
            self.__Head = (self.__Head).next;
            del((self.__Head).prev);
            (self.__Tail).next = self.__Head;
            (self.__Head).prev = self.__Tail;

    def DeleteLast(self) :
        if((self.__Head == None)and(self.__Tail == None)) :
            return;
        elif(self.__Head == self.__Tail) :
            del(self.__Head);
            self.__Head = None;
            self.__Tail = None;
        else :
            self.__Tail = (self.__Tail).prev;
            del((self.__Tail).next);
            (self.__Tail).next = self.__Head;
            (self.__Head).prev = self.__Tail;

    def DeleteAtPos(self,ipos) :
        i = 0;
        isize = 0;
        temp = self.__Head;
        temp2 = None;

        isize = self.Count();

        if((ipos < 1)or(ipos > isize)) :
            return;
        elif(ipos == 1) :
            self.DeleteFirst();
        elif(ipos == isize) :
            self.DeleteLast();
        else :
            for i in range(1,ipos - 1) :
                temp = temp.next;
            
            temp2 = temp.next;
            temp.next = temp2.next;
            temp2.next.prev = temp;
            del(temp2);
            temp2 = None;
 
    def Count(self) :
        Head = self.__Head;
        Tail = self.__Tail;
        icnt = 0;

        if((Head == None)and(Tail == None)) :
            return;
        else :
            while(Head != Tail) :
                icnt = icnt + 1;                
                Head = Head.next;
            icnt = icnt + 1;

            return icnt;

def main() :
    ino = 0;
    iret = 0;
    ielem = 0;
    ipos = 0;

    print("Jay Ganesh....");

    print("------------------------------Doubly Circular Linked List Demonstration--------------------------");

    dobj = DoublyCLL();

    while(True) :
        print("Select Any One Operation From Below List of Operations : ");
        print("1. Insert Value At First Position.");
        print("2. Insert Value At Last Position.");
        print("3. Insert Value At Given Position.");
        print("4. Display Elements of Linked List.");
        print("5. Count No of Elements Present in Linked List.");
        print("6. Delete Element From First Position.");
        print("7. Delete Element From Last Position.");
        print("8. Delete Element From Given Position.");
        print("0. Exit.");

        print("Enter Your Choice : ");
        ino = int(input());

        if(ino == 1) :
            print("Enter One Value to Insert : ",end = "");
            ielem = int(input());

            dobj.InsertFirst(ielem);
        elif(ino == 2) :
            print("Enter One Value to Insert : ",end = "");
            ielem = int(input());

            dobj.InsertLast(ielem);
        elif(ino == 3) :
            print("Enter One Value to Insert : ",end = "");
            ielem = int(input());

            print("Enter Position To Insert Element At : ",end = "");
            ipos = int(input());

            dobj.InsertAtPos(ielem,ipos);

            pass;
        elif(ino == 4) :
            dobj.Display();
        elif(ino == 5) :
            iret = dobj.Count();

            print("Count of Element Present in Linked List is : ",iret);
        elif(ino == 6) :
            dobj.DeleteFirst();
        elif(ino == 7) :
            dobj.DeleteLast();
        elif(ino == 8) :
            print("Enter Position To Delete Element From : ",end = "");
            ipos = int(input());

            dobj.DeleteAtPos(ipos);
        elif(ino == 0) :
            return;
        else :
            print("Invalid Option Number,Please Enter valid Option Number");

            return;

if __name__ == "__main__" :
    main();